blog/content/posts/2025-xx-xx-using-server-instructions.md

```diff
@@ -0,0 +1,220 @@
+
+One key to good instructions is focusing on **what tools/resources don't convey**:
+
+1. **Capture cross-feature relationships**:
+    
+    ```json
+    {
+      "instructions": "Always call 'authenticate' before any 'fetch_*' tools. The 'cache_clear' tool invalidates all 'fetch_*' results."
+    }
+    ```
+    
+2. **Document operational patterns**:
+    
+    ```json
+    {
+      "instructions": "For best performance: 1) Use 'batch_fetch' for multiple items, 2) Check 'rate_limit_status' before bulk operations, 3) Results are cached for 5 minutes."
+    }
+    ```
+    
+3. **Specify constraints and limitations**:
+    
+    ```json
+    {
+      "instructions": "File operations limited to workspace directory. Binary files over 10MB will be rejected. Rate limit: 100 requests/minute across all tools."
+    }
+    ```
+    
+
+### Anti-Patterns to Avoid
+
+❌ **Don't repeat tool descriptions**:
+
+```json
+// Bad - duplicates what's in tool.description
+"instructions": "The search tool searches for files. The read tool reads files."
+
+// Good - adds relationship context
+"instructions": "Use 'search' before 'read' to validate file paths. Search results expire after 10 minutes."
+```
+
+❌ **Don't include marketing or superiority claims**:
+
+```json
+// Bad
+"instructions": "This is the best server for all your needs! Superior to other servers!"
+
+// Good
+"instructions": "Specialized for Python AST analysis. Not suitable for binary file processing."
+```
+
+❌ **Don't write a manual**:
+
+```json
+// Bad - too long and detailed
+"instructions": "This server provides comprehensive functionality for... [500 words]"
+
+// Good - concise and actionable
+"instructions": "GitHub integration server. Workflow: 1) 'auth_github', 2) 'list_repos', 3) 'clone_repo'. API rate limits apply - check 'rate_status' before bulk operations."
+```
+
+### What Server Instructions Can't Do:
+
+- **Guarantee certain behavior:** As with any text you give to an LLM, your instructions aren't going to be followed the same way 100% of the time.  Anything you ask a model to do is like rolling a dice  The reliability of any instructions will vary based on randomness, sampling parameters, model, client implementation, other servers/tools at play, and many other variables.
+	- Given the above, don't rely on instructions for any critical 'must-do' actions that need to happen in conjunction with other actions, especially in security or privacy domains.  These are better implemented as deterministic rules or hooks.
+- **Make up for suboptimal tool design:** Tool descriptions and other aspects of interface design for agents are still going to make or break how well LLMs can use your server, when they need to take an action.
+
+## Implementing Server Instructions: For Client Developers
+
+If you're a client developer, your job is a more complex, since it involves deciding how to incorporate instructions into what ultimately gets passed to the model.
```

Comment:
```suggestion
If you're a client developer, your job is more complex, since it involves deciding how to incorporate instructions into what ultimately gets passed to the model.
```